### PR TITLE
feat(delivery): expose message delivery state in the UI [#862]

### PR DESCRIPTION
**Pull request by lsm**
Phase 4 of message-delivery reliability. Surfaces the durable `send_status` lifecycle (`deferred`/`enqueued`/`submitted`/`consumed`/`failed`) to users as explicit delivery badges (`queued`/`processing`/`retrying`/`delivered`/`failed`) instead of inferring state from system/init messages.

Shared `MessageDeliveryStatus` type + `sendStatusToDeliveryStatus` mapper drive both the `messages.bySession` (ordinary chat) and `spaceTaskMessages` (Space thread) feeds. The feeds now show `deferred`/`enqueued`/`submitted` user rows and map `send_status` + the active `message_delivery` job `retry_count` (for `retrying`); daemon-side reads keep the old `consumed`/`failed` filter so prompt context and rewind are unaffected. The widened feed also gives a natural optimistic echo — the persisted `enqueued` row appears within ms of send via the reactive-DB flush and transitions in place by stable row id (no web-side optimistic insert, no duplicate bubble on retry/transition).

`SDKUserMessage` renders a badge for queued/processing/retrying/failed and hides it for delivered (avoid noise); Space threads expand from delivered/failed to the full lifecycle. Steer-consumed propagation (item 11) is satisfied by the existing synchronous consumed-flip at `onSent` plus the now-visible feed. The Space wake/recovery audit (items 6/7) found no demonstrably redundant mechanisms — the agent-level duplicate was already removed in #861, and the surviving Space recovery paths (nudge/nag/terminal-error-continue/restart-notice/external-event requeue) each own a distinct failure mode — so none were removed.

Verified: unit + reactive-pipeline integration tests for the feed mapping, retry signal, and no-duplicate transitions; web component tests for the badges; daemon+web typecheck/lint/knip; `make build`; dev-server boot smoke test. `bun run check` has one pre-existing `check:test-quality` failure in an untouched provider test (`provider-registry.test.ts`), unrelated to this work. Full interactive browser verification of a Space handoff + controlled stale-delivery recovery is left for QA/review (needs a live Space + credentials).

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (HyperNeo)

> **Model:** Claude (glm-5.1) | **Client:** HyperNeo | **Provider:** Anthropic

Phase 4 of message-delivery reliability: surfaces the durable `send_status` lifecycle as explicit delivery badges (`queued`/`processing`/`retrying`/`delivered`/`failed`) in the chat and Space feeds, instead of inferring state from system/init messages. A shared `MessageDeliveryStatus` type + `sendStatusToDeliveryStatus` mapper drive both feeds; the retry signal comes from an `EXISTS (... message_delivery job ... retry_count > 0)` subquery; daemon-side prompt-context/rewind reads keep the old consumed/failed filter.

### Verdict: Request changes — no P0/P1, a few P2s worth addressing (one consistency gap + test-coverage gaps for the new reactive path)

The core implementation is correct. I verified the highest-risk concern directly: **the "retrying" badge is genuinely reactive**. The `messages.bySession` SQL now references `job_queue` (via the EXISTS subquery), `extractTables` (`live-query.ts:134`) captures `job_queue` as a feed dependency, and `messageDeliveryProcessor.setChangeNotifier(...)` in `app.ts:379-381` routes every job transition to `reactiveDb.notifyChange('job_queue')`. So a `retry_count` bump re-evaluates the feed. The partial index `idx_message_delivery_session_active` makes the correlated EXISTS a bounded seek (N × (seek + K×json_extract), K≈1). The daemon-side `SDKMessageRepository` reads still apply the consumed/failed filter (`sdk-message-repository.ts:693,756,781,885,...`), so no `deferred`/`enqueued` row leaks into prompt context — no double-send. The `sendStatus` wire field had exactly one web reader (`SDKUserMessage.tsx`), which this PR updates.

### P2 — Dual wire shape: `getSDKMessages` HTTP path still emits the old `sendStatus` field

`sdk-message-repository.ts:848-850` (the `messages` RPC / initial-load / pagination / export path, **not in this diff**) still attaches `extra.sendStatus = 'failed'`, while the `messages.bySession` LiveQuery feed now emits `deliveryStatus` and no longer emits `sendStatus`. Two different wire shapes for "SDK messages" depending on transport. No web consumer reads either field from the RPC path today (the LiveQuery feed is the sole transcript source), so there's **no functional break now** — but it's a latent trap for a future consumer that reads `deliveryStatus` from an initial-load payload and finds it absent. Either migrate both paths to `deliveryStatus` (run `sendStatusToDeliveryStatus` in the mapper) or add a one-line note to the docs/PR that the split is intentional. (Inline comment below.)

### P2 — Test gap: no test verifies the new `job_queue` reactive path

The implementation is reactive, but no test proves it. `live-query-delivery-status.test.ts` mutates only `sdk_messages` (`UPDATE send_status` + `notifyChange('sdk_messages')`) — it never touches `job_queue`. `live-query-messages.test.ts` exercises the static SQL mapping only (insert then re-query). A regression that, say, drops `job_queue` from `extractTables` capture or unwires `messageDeliveryProcessor.setChangeNotifier` would pass all current tests while silently making the "retrying" badge non-reactive. Add one test: subscribe to `messages.bySession`, insert a `job_queue` row with `retry_count > 0`, assert an UPDATED delta arrives on the matching message row. This is the single most valuable test to add — it locks in the one non-obvious architectural property of this change.

### P2 — Test gap: `spaceTaskMessages.byTask(.compact)` retry signal has no coverage

The same EXISTS retry subquery + widened `deliveryState` mapping was added to both `spaceTaskMessages` CTEs (`live-query-handlers.ts:~1952`, `~2264`), but `live-query-messages.test.ts` tests only `messages.bySession` (the helper hardcodes that named query). The Space feed's retry signal and full-lifecycle pass-through are untested.

### P3 — `retrying` web state and tooltip copy untested

`SDKUserMessage.test.tsx` covers `failed`/`queued`/`processing`/`delivered`/absent but omits `retrying` — the most novel state, and the one that shares amber styling with `queued` (disambiguated only by label "retrying" vs "queued"). A swap would pass silently. (Tooltip copy is never asserted either, but the Tooltip requires hover+delay and synchronous render can't reach it — acceptable to leave.)

### P3 — SQL CASE/EXISTS triplication

The same delivery-state logic is hand-written in three places: two byte-identical inline CASE blocks (`~1955`, `~2267`) and a structurally different EXISTS-column + JS-mapper form in `messages.bySession` (`~3300`). Consistent with this file's existing inline-duplication convention, and the JS `sendStatusToDeliveryStatus` is the true single-source-of-truth for the *logic* — but the two Space CASE blocks are a redundant reimplementation that must be kept in sync by hand. Not a blocker; clean extraction candidate if the file is ever refactored.

### P3 — `job_queue` change notifications fire without session scope

`notifyChange('job_queue')` (`job-queue-processor.ts:196`) carries no scope, so one session's job transition re-evaluates every open `messages.bySession`/`spaceTaskMessages` subscription. Acceptable today (debounced; EXISTS is index-bounded; few concurrent deliveries per session), but worth knowing if multi-session concurrency grows.

### Recommendation
Recommendation: REQUEST_CHANGES — the P2s are all worth a small follow-up (one mapper call for wire-shape parity, two/three tests to lock the reactive + Space paths). None are blocking for correctness; the implementation is sound. Deferring the final runtime/browser verification (task item 9) and the `getSDKMessages` parity call to QA.

---
Recommendation: REQUEST_CHANGES

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 1]
```diff
@@ -2240,7 +2264,24 @@ sdk_rows AS (
     sm.message_type AS messageType,
     sm.sdk_message AS content,
     sm.origin AS origin,
-    CASE WHEN sm.message_type = 'user' AND sm.send_status = 'failed' THEN 'failed' WHEN sm.message_type = 'user' THEN 'delivered' ELSE NULL END AS deliveryState,
+    CASE
+      WHEN sm.message_type != 'user' THEN NULL
+      WHEN COALESCE(sm.send_status, 'consumed') = 'failed' THEN 'failed'
+      WHEN COALESCE(sm.send_status, 'consumed') = 'consumed' THEN 'delivered'
+      WHEN COALESCE(sm.send_status, 'consumed') IN ('deferred', 'enqueued', 'submitted')
+           AND EXISTS (
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:2285]: Fixed in bf25affd9: the compact feed's recent-turn cutoff now includes active nonterminal user rows independently of `conversation_turn_index >= rt.minTurn`, so a queued prompt to a dormant agent (old turn index) is no longer dropped. Covered by 'includes a queued prompt to a dormant session even when its turn is older than the recent-turn cutoff'.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 2]
```diff
@@ -3251,11 +3291,30 @@ WITH top_level AS (
...
   FROM sdk_messages
   WHERE session_id = ?1
     AND parent_tool_use_id IS NULL
-    AND (message_type != 'user' OR COALESCE(send_status, 'consumed') IN ('consumed', 'failed'))
+    -- Task #862: surface ALL user-message delivery states (deferred / enqueued
+    -- / submitted / consumed / failed) so the UI can badge queued / processing
+    -- / retrying explicitly instead of inferring from system/init. Daemon-side
+    -- reads (SDKMessageRepository.getSDKMessages) keep the consumed/failed
+    -- visibility filter, so prompt context / rewind are unaffected; only the
+    -- web transcript feed widens.
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:3345]: Fixed in bf25affd9: SessionStore._applyMessagesDelta now re-sorts by (timestamp, rowid) after an updated row whose delivery-status change moved its timestamp, so the prompt reorders without needing a reconnect snapshot. Covered by 're-sorts updated rows when a delivery-status update moves their timestamp'.

File: packages/web/src/components/sdk/SDKUserMessage.tsx [hunk 3]
```diff
@@ -351,13 +388,8 @@ export function SDKUserMessage({
...
-      {message.sendStatus === 'failed' && (
-        <Tooltip
-          content="Message was not delivered — the server crashed before Claude responded"
-          position="left"
-        >
-          <span class="text-xs px-2 py-0.5 bg-red-500/20 text-red-400 rounded">not delivered</span>
-        </Tooltip>
+      {message.deliveryStatus && message.deliveryStatus !== 'delivered' && (
+        <UserMessageDeliveryBadge status={message.deliveryStatus} />
       )}
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/web/src/components/sdk/SDKUserMessage.tsx:399]: Fixed in 24e880767 (before this review): SDKMessageRepository._getSDKMessagesImpl now emits deliveryStatus via sendStatusToDeliveryStatus, so paginated (message.sdkMessages) failed messages carry the badge the renderer reads. No legacy fallback needed.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:3345]: Fixed in bf25affd9: runPostSaveSideEffects (the outbox's only post-commit hook) now always notifies sdk_messages (scoped to the session), so the widened feed is re-evaluated on the initial enqueued commit and the queued row surfaces immediately. Covered by the outbox + reactive-pipeline tests.

File: packages/web/src/components/sdk/SDKUserMessage.tsx [hunk 4]
```diff
@@ -351,13 +388,8 @@ export function SDKUserMessage({
...
 
-      {message.sendStatus === 'failed' && (
-        <Tooltip
-          content="Message was not delivered — the server crashed before Claude responded"
-          position="left"
-        >
-          <span class="text-xs px-2 py-0.5 bg-red-500/20 text-red-400 rounded">not delivered</span>
-        </Tooltip>
+      {message.deliveryStatus && message.deliveryStatus !== 'delivered' && (
+        <UserMessageDeliveryBadge status={message.deliveryStatus} />
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/web/src/components/sdk/SDKUserMessage.tsx:398]: Fixed in bf25affd9: SDKUserMessage hides the rewind button for queued/processing/retrying delivery (terminal delivered/failed — or a legacy row without deliveryStatus — still shows it), so a rewind can no longer delete a row whose active message_delivery job is still executing. Covered by the 'Rewind gating' describe.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:2285]: Fixed in bf25affd9: user_entries in the active-turn roster SQL excludes send_status='deferred', so a deferred prompt stays in the main thread but is not shown as in-turn input. Covered by 'keeps an in-turn prompt in the active roster but excludes a deferred one'.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 5]
```diff
@@ -3251,11 +3291,30 @@ WITH top_level AS (
...
+    -- (retry_count > 0). Scoped to this session's active jobs (few) via the
+    -- partial index idx_message_delivery_session_active; one bounded lookup.
+    EXISTS (
+      SELECT 1
+      FROM job_queue jq
+      WHERE jq.queue = 'message_delivery'
+        AND jq.status IN ('pending', 'processing')
+        AND json_extract(jq.payload, '$.sessionId') = ?1
+        AND json_extract(jq.payload, '$.messageUuid') = sdk_messages.sdk_uuid
+        AND jq.retry_count > 0
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:3335]: Fixed in bf25affd9: revokePendingDelivery('defer') now notifies sdk_messages + job_queue (via a Database.notifyChange bridge) after deferEnqueuedUserMessage + cancelDelivery, so deliveryRetry drops 1->0 and the transcript stops showing 'retrying'.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 6]
```diff
@@ -3257,11 +3293,30 @@ WITH top_level AS (
...
-    rowid
+    rowid,
+    -- Active-delivery retry signal for the "retrying" UI state (task #862):
+    -- true when a pending/processing message_delivery job for this session
+    -- references the row's canonical uuid AND has already been retried
+    -- (retry_count > 0). Scoped to this session's active jobs (few) via the
+    -- partial index idx_message_delivery_session_active; one bounded lookup.
+    EXISTS (
+      SELECT 1
+      FROM job_queue jq
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:3330]: Fixed in 72d807d8e: JobQueueProcessor now passes a TableChangeScope (from the job payload sessionId/taskId) to its change notifier; app.ts wires the generic non-delivery processor to a no-op for job_queue and the message_delivery processor notifies session-scoped. The feeds depend on job_queue only for the message_delivery retry signal, so unrelated schedule/cleanup/polling/workflow jobs no longer re-run every open feed. Covered by the processor scoped-notify tests.

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 7]
```diff
@@ -1969,8 +1990,7 @@ sdk_rows_raw AS (
    AND sne.rn = 1
   LEFT JOIN space_agents sa
     ON sa.id = COALESCE(sne.agent_id, json_extract(s_kind.metadata, '$.promptProvenance.agentId'))
-  WHERE (sm.message_type != 'user' OR COALESCE(sm.send_status, 'consumed') IN ('consumed', 'failed'))
-    AND (
+  WHERE (
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:1993]: Fixed in 72d807d8e: the full feed's sdk_rows_with_pos only sets thisUserRowPos (the turn sentinel) for SETTLED (consumed/failed) user rows, so a pending prompt is emitted for its badge but does not become the turnId subsequent assistant rows inherit. Covered by 'a pending user row does not become the turnId for subsequent assistant rows'.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Fable 5 (Anthropic)

> **Model:** Fable 5 | **Client:** HyperNeo | **Provider:** Anthropic

Fresh review at head `72d807d8e` (own-PR fallback: COMMENT review carrying `Recommendation: REQUEST_CHANGES`). I re-derived the reactive dependency wiring, the interrupt path, and the sibling notify the PR already applied — independent of the prior rounds.

### What's solid

The core design is right and well-verified:
- **`messageUuid` payload key matches the writer.** Every `message_delivery` job is enqueued from `deliverMessage` with a `messageUuid` field (camelCase), and the `deliveryRetry` EXISTS subqueries (`live-query-handlers.ts:1968, 2291, 3334, 3412`) read `json_extract(jq.payload, '$.messageUuid')` against `sm.sdk_uuid` — exact match, no silent miss. The partial index `idx_message_delivery_session_active` (migration 186) backs each EXISTS as a bounded lookup.
- **Non-user rows are correctly gated.** `mapMessageRow` only attaches `deliveryStatus` when `parsed.type === 'user'` (`live-query-handlers.ts:3480`), and the Space feeds use an inline `CASE WHEN sm.message_type != 'user' THEN NULL`. A NULL `send_status` (assistant/system/result rows) coalesces to `'consumed'` but never becomes a spurious `delivered` badge.
- **Turn-anchoring change is consistent.** `thisUserRowPos` now only fires for terminal (delivered/failed) user rows (`live-query-handlers.ts:2042-2048`), aligning with `saveUserMessageCore` withholding `isConversationAnchor` until consumed/failed. A `submitted` row that never settles inherits the prior terminal turn's `turnId` (via the `MAX(thisUserRowPos)` window) — trailing assistant rows don't get a null/gap turn key. No grouping regression.
- **Daemon SDK-context path is unaffected.** `getSDKMessages` keeps the old `('consumed','failed')` visibility filter (`sdk-message-repository.ts:~754`), so prompt context and rewind still see only settled user rows. The widening lives solely in the two LiveQuery feeds.
- **No schema change.** All five `send_status` states pre-date this PR (migration 183 widened the CHECK to `submitted`).
- **Re-sort is stable and cursor-consistent.** `sortByTimestampRowid` matches the daemon's `ORDER BY timestamp ASC, rowid ASC`; necessary because `updateMessageStatus` re-stamps the timestamp on settle.

### P1 — Interrupt path doesn't notify the delivery feeds (stuck badge)

**Confirmed true** (this matches the open Codex thread; I verified it independently).

`messages.bySession` and `spaceTaskMessages` auto-depend on **both** `sdk_messages` and `job_queue`: `extractTables` (`live-query.ts:134`) derives dependencies from the SQL text and picks up `FROM job_queue` inside the `deliveryRetry` EXISTS subquery. The new reactive test even asserts this ("a job_queue retry transition re-evaluates the feed"). So any raw-db mutation to those two tables **must** call `notifyChange(table, {sessionId})` or the feed goes stale.

The PR applied that notify to the sibling path — `revokePendingDelivery` (`agent-session.ts:956-957`, the Move-to-Next RPC), with an explicit comment explaining exactly this:

```ts
this.db.notifyChange?.('sdk_messages', { sessionId: this.session.id });
this.db.notifyChange?.('job_queue', { sessionId: this.session.id });
```

But the **interrupt chokepoint** — which performs the same two writes in bulk — was missed. `InterruptHandler.handleInterrupt` (`interrupt-handler.ts:82-108`, inside `if (!opts.preserveDeliveryJobs)`) calls:
- `cancelForSessionWithMessages` → raw `DELETE FROM job_queue` (`job-queue-repository.ts:279-302`), and
- `markDeliveryFailedByUuid` → `updateMessageStatus` → raw `UPDATE sdk_messages SET send_status='failed'` (`sdk-message-repository.ts:1430-1507`).

`grep` confirms **zero** `notifyChange`/`reactiveDb` references in `interrupt-handler.ts`; `JobQueueRepository` has no reactive integration at all; and `updateMessageStatus` only notifies `sessions` (and only when the badge counter changes — a deferred/enqueued/submitted→failed flip typically doesn't). Net effect: after a user **interrupts while a message is still pending/enqueued (or a steer job is in flight)**, the already-rendered `queued`/`retrying` badge stays stuck until an unrelated write or a reconnect re-evaluates the query. That is precisely the scenario item 11 (steer-consumed) and the "no second user message needed" success criterion care about.

**Fix:** mirror the two notify calls inside the `if (!opts?.preserveDeliveryJobs)` block, after the `withSessionLock` callback closes (~`interrupt-handler.ts:107`). `this.ctx.db` is the `Database` facade carrying the new `notifyChange`, so the call form is identical to `agent-session.ts:956-957`:

```ts
      }); // end withSessionLock
      this.ctx.db.notifyChange?.('sdk_messages', { sessionId: session.id });
      this.ctx.db.notifyChange?.('job_queue', { sessionId: session.id });
    }
```

### P2 — Stale `sendStatus?: string` type on `getSDKMessages`

The runtime now emits `deliveryStatus` (`sdk-message-repository.ts:853-857`), but the declared return type at `sdk-message-repository.ts:662, 747, 930` still advertises `sendStatus?: string`. No current consumer reads `.sendStatus`, so this is a type-level lie rather than a runtime bug — but it'll mislead the next caller. Update the return type to `deliveryStatus?: MessageDeliveryStatus` and drop the dead `sendStatus`.

### P2 — Notify-correctness paths have no test

The new reactive test (`live-query-delivery-status.test.ts`) mutates `sdk_messages`/`job_queue` directly and proves the feed re-evaluates — but it does **not** exercise either *production* notify path (`revokePendingDelivery` or `handleInterrupt`). So the P1 fix above (and the already-landed sibling) are statically trusted, not regression-guarded. When the interrupt notify is added, please add a daemon test that drives `handleInterrupt` against a pending/enqueued message and asserts the feed re-evaluates without an unrelated write.

### P2 — Generic job-processor notifier silenced to a no-op

`app.ts:381` sets the generic processor's notifier to `() => {}`. The comment is accurate today (transcript feeds touch `job_queue` only for the delivery retry signal), and I confirmed no other live query subscribes to `job_queue` beyond the per-row EXISTS subquery. But the coupling is now implicit — a future `job_queue`-subscribing feed would silently lose reactivity with no compile-time signal. Consider threading a scope that session-scoped feeds ignore, or at least documenting the invariant where `job_queue` live queries get registered. Low urgency; no current consumer breaks.

### P3 — `retrying` and `queued` share identical styling

`DeliveryStateBadge.tsx:15-17` gives `retrying` and `queued` the same amber classes. They're distinguishable by label, so this is fine — but `retrying` implies a stall and could read as more urgent. Pure preference.

### Verdict

One P1 (stuck badge on interrupt — the success criterion's exact scenario) and three non-blocking P2/P3s. The P1 is a small, well-scoped fix (two lines mirroring a pattern already in this PR) plus a test. Recommend addressing the P1 before this lands.

Recommendation: REQUEST_CHANGES

File: packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts [hunk 8]
```diff
@@ -3257,11 +3319,30 @@ WITH top_level AS (
...
   FROM sdk_messages
   WHERE session_id = ?1
     AND parent_tool_use_id IS NULL
-    AND (message_type != 'user' OR COALESCE(send_status, 'consumed') IN ('consumed', 'failed'))
+    -- Task #862: surface ALL user-message delivery states (deferred / enqueued
+    -- / submitted / consumed / failed) so the UI can badge queued / processing
+    -- / retrying explicitly instead of inferring from system/init. Daemon-side
+    -- reads (SDKMessageRepository.getSDKMessages) keep the consumed/failed
+    -- visibility filter, so prompt context / rewind are unaffected; only the
+    -- web transcript feed widens.
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/live-query-handlers.ts:3345]: Fixed in d3a14d5b9. `InterruptHandler.handleInterrupt` (the cancel-everything path, i.e. `if (!opts?.preserveDeliveryJobs)`) now calls `notifyChange('sdk_messages', { sessionId })` + `notifyChange('job_queue', { sessionId })` on the Database facade after the `withSessionLock` block closes, so cancelling a pending delivery job + flipping the enqueued row to failed drops the queued badge immediately (no stuck badge until an unrelated write/reconnect). Mirrors the `revokePendingDelivery` notify. Covered by 'notifies the delivery feeds (sdk_messages + job_queue) after cancelling deliveries' in interrupt-handler.test.ts.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Fable 5 (Anthropic)

> **Model:** Fable 5 | **Client:** HyperNeo | **Provider:** Anthropic

Round-3 fresh review at head `d3a14d5b9`. Own-PR fallback: COMMENT review carrying `Recommendation: APPROVE`. I re-derived the reactive scoping, the interrupt/revoke notify paths, the wire-shape rename, and the SQL CASE/EXISTS independently, and fanned out dimension sub-agents for correctness, impact, security, tests/performance, craft, and UX. All six converged clean.

### All prior findings addressed ✅

I verified each prior-round item against the actual code:

- **P1 interrupt notify** — `interrupt-handler.ts:108-115` now calls `notifyChange('sdk_messages'|'job_queue', {sessionId})` after the `withSessionLock` block closes, inside the `if (!opts?.preserveDeliveryJobs)` guard, before the idle/interrupted early-return. Both raw writes (`cancelForSessionWithMessages` → raw DELETE; `markDeliveryFailedByUuid` → raw UPDATE) bypass the reactive proxy's `METHOD_TABLE_MAP`, so the explicit notify is necessary, not redundant. Covered by `interrupt-handler.test.ts` asserting both tables + session scope on the real handler.
- **P2 sendStatus type parity** — `sdk-message-repository.ts:660/745/928` now return `deliveryStatus?: MessageDeliveryStatus`. `mapMessageRow` emits `deliveryStatus` via the shared `sendStatusToDeliveryStatus`. Daemon-side reads (`getSDKMessagesImpl`) keep the `('consumed','failed')` visibility filter — prompt context and rewind unaffected. (Two forwarder signatures still declare the stale type — see P2 below.)
- **P2 notify-path tests** — interrupt-handler + agent-session tests drive the production handlers and assert both notifies; the reactive-pipeline test covers updated-not-duplicated + reconnect + job_queue dependency.
- **P2 generic-notifier invariant** — documented in `app.ts:137-146` with a clear migration note; verified no non-delivery feed joins `job_queue`.
- **P3 retrying styling** — `DeliveryStateBadge.tsx` now reads orange (`retrying`) vs amber (`queued`).

### Scoping is real and correct

I traced the whole chain because it's the load-bearing design: `job_queue` becomes a dependency of every transcript/task feed (via the retrying `EXISTS`), so a naive notify would re-run all open feeds on every unrelated job. The fix threads a `TableChangeScope` from the job payload (`scopeFromJob`) → `messageDeliveryProcessor.setChangeNotifier((table, scope) => reactiveDb.notifyChange(table, scope))`, while the generic processor is a no-op. `LiveQueryEngine.onTableChange` (`live-query.ts:417`) skips re-evaluation when `scopeFilter(scope)` is false; `messages.bySession`'s filter matches `scope.sessionId === targetSessionId`, and `spaceTaskMessages`'s `buildTaskScopeFilter` dynamically links a `{sessionId}` scope to its task. So a session-scoped delivery notify reaches exactly the right feed and skips the rest. The `messageUuid` EXISTS join key (camelCase payload) matches the writer (`message-delivery.ts`), `sdk_uuid` is reliably populated, and `NULL = x` is safely false in SQLite. No N+1 — the partial index `idx_message_delivery_session_active` bounds the EXISTS to ≤ a few active jobs per session.

### Findings (none blocking)

**P2 — Missing negative-scope test.** The scoping invariant's whole point — a scoped `job_queue` notify for session A does *not* re-evaluate session B's feed — is asserted only in the `app.ts` comment, never in a test. `job-queue-processor.test.ts` checks the scope is *passed*; `live-query-delivery-status.test.ts` checks an *unscoped* notify re-evaluates. Nothing subscribes two sessions and asserts the scoped notify is filtered. A future refactor of `scopeFilter`/`onTableChange` could silently regress this with no test failure. Worth a small two-session reactive test.

**P2 — Stale `sendStatus?: string` on two forwarder signatures.** The repo's `getSDKMessages` type was fixed, but `storage/index.ts:228` (`Database.getSDKMessages` facade) and `agent-session.ts:1657` still declare `sendStatus?: string`. TS is structural, no caller breaks (the property is simply absent at runtime), and no consumer reads `.sendStatus` — so this is a type-level lie, not a runtime bug. Finish the rename for consistency.

**P2 — No SQL↔TS sync guard.** The `deliveryState` CASE is duplicated verbatim across the full and compact feeds and must stay byte-for-byte in sync with the TS `sendStatusToDeliveryStatus` mapper declared the source of truth. There's no test that asserts parity (the TS mapper has no unit tests). A one-line comment in each CASE pointing to the canonical table, or a focused mapper unit test, would cheaply close this.

**P1 (a11y, pre-existing surface) — Badge tooltip is mouse-only; badge lacks ARIA semantics.** `SDKUserMessage.tsx:107-109` wraps the badge in `Tooltip`, which only binds `onMouseEnter`/`onMouseLeave` and the badge is a non-focusable `<span>`. The explanatory copy ("Delivery stalled — retrying", "Message was not delivered — the server crashed before Claude responded") never reaches keyboard/screen-reader users — and for `failed`, the *why* is load-bearing. The badge has no `role`/`aria-label`. This is a pre-existing Tooltip limitation this PR makes matter for a user-facing status affordance. Cheapest fix: `role="alert"` (failed) / `role="status"` (non-terminal) + `aria-label` with the tooltip copy on the span. Not a regression introduced here, so not blocking this PR — but worth a follow-up since the badges are new user-facing surfaces.

**P3 — CASE `ELSE 'queued'` vs TS `null` for unknown send_status.** A hypothetical future/legacy `send_status` value falls through to `'queued'` in SQL but `null` (no badge) in the TS path. No unrecognized value can reach there today (every write site uses the enum), so no live repro — align the fallbacks if you add a sixth status.

**P3 — UX copy.** `processing` label reads "sending" (accurate pre-consume, slightly odd once the assistant is mid-reply); `failed` tooltip asserts "the server crashed" but `failed` can also mean exhausted retries. Minor; not blocking.

**P3 — Facade `notifyChange` scope type narrower than `TableChangeScope`.** `storage/index.ts:142` hand-rolls `{sessionId?; taskId?}` instead of importing `TableChangeScope`. Harmless today (all callers pass `{sessionId}`); import the shared type to track the canonical contract.

### Verdict

Zero correctness, security, or performance blockers. Every prior-round item is addressed and verified against the code. The scoping design is sound and correctly wired. Remaining items are polish (two stale forwarder types, a missing negative-scope test, a TS↔SQL sync note) and a pre-existing a11y gap on the new badge surface — none block merge. Recommend QA cover the browser-level Space handoff + stale-delivery recovery (deferred per the PR) and the badge a11y path.

Recommendation: APPROVE

**Review by lsm — COMMENTED** [review]
**QA Re-Validation (post-approval CI fix)**

Head c1a591db1 is a single commit wrapping two `deliveryStatus` return-type declarations multiline — a pure Biome formatting change (no logic or behavior change). All 33 CI checks pass (All Tests Pass ✅). PR is MERGEABLE/CLEAN.

Recommendation: APPROVE
